Add GPU operator chart

Ajarmar opened the pull request:
<!-- Choose your PR title carefully as it will be used as the entry in the changelog! -->

<!-- markdownlint-disable MD041 -->
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       <!-- This PR adds a new feature -->
- [ ] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

_Optional_: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change   <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change     <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security       <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr](set-me\) <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

<!-- Additional information with kind/admin-change
### Platform Administrator notice
...
-->

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

<!-- Add description of the change -->
Adds NVIDIA's GPU operator helm chart.

I tested setting it up in an environment, but I haven't tested using it with an actual GPU node yet; just wanted to get the PR up first to start getting reviews.

<!-- Add all issues that are fixed by this PR, use "Part of" instead of "Fixes" if you want to keep issues open. -->
- Fixes elastisys/ck8s-issue-tracker#415

#### Information to reviewers

<!--
Any additional information reviews should know.

How to run / how to test.

Include screenshots if applicable to help explain these changes.
--->

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [x] Proper commit message prefix on all commits
    <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to applications running in all clusters
    - apps sc: changes to applications running in service clusters
    - apps wc: changes to applications running in workload clusters
    - bin: changes to management binaries
    - config: changes to configuration
    - docs: changes to documentation
    - release: release related
    - scripts: changes to scripts
    - tests: changes to tests
    --->
- Change checks:
    - [x] The change is transparent
    - [ ] The change is disruptive
    - [x] The change requires no migration steps
    - [ ] The change requires migration steps
    - [ ] The change updates CRDs
    - [x] The change updates the config _and_ the schema
- Documentation checks:
    - [x] The [public documentation](https://github.com/elastisys/welkin) required no updates
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required an update - [link to change](set-me\)
- Metrics checks:
    - [ ] The metrics are still exposed and present in Grafana after the change
    - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts required no updates)
    - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts required an update)
- Logs checks:
    - [ ] The logs do not show any errors after the change
- PodSecurityPolicy checks:
    - [ ] Any changed Pod is covered by Kubernetes Pod Security Standards
    - [ ] Any changed Pod is covered by Gatekeeper Pod Security Policies
    - [ ] The change does not cause any Pods to be blocked by Pod Security Standards or Policies
- NetworkPolicy checks:
    - [ ] Any changed Pod is covered by Network Policies
    - [ ] The change does not cause any dropped packets in the NetworkPolicy Dashboard
- Audit checks:
    - [ ] The change does not cause any unnecessary Kubernetes audit events
    - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
    - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
    - [ ] The bug fix is covered by regression tests

Xartos commented on config/schemas/config.yaml:
```diff
@@ -7409,6 +7409,53 @@ properties:
+              tolerations:
+                $ref: '#/$defs/kubernetesTolerations'
+              affinity:
+                $ref: '#/$defs/io.k8s.api.core.v1.Affinity'
```
Question: You missed the daemonset part here right?

Ajarmar replied:
Yep, I forgot it. Pushed it now.

Xartos commented on config/schemas/config.yaml:
```diff
@@ -7409,6 +7409,53 @@ properties:
+            properties:
+              resources:
+                $ref: '#/$defs/kubernetesResourceRequirements'
+              tolerations:
+                $ref: '#/$defs/kubernetesTolerations'
```
Question: Do we need to specify the defaults from the config here? i.e. the `nvidia.com/gpu:NoSchedule` for this one

Ajarmar replied:
Hmm, that's a good point. But I couldn't find any tolerations in the config that we define defaults for, even if they have defaults in the config (see e.g. `falcoExporter`: defaults [in the config](https://github.com/elastisys/compliantkubernetes-apps/blob/v0.43.0/config/common-config.yaml#L201), but [not in the schema](https://github.com/elastisys/compliantkubernetes-apps/blob/v0.43.0/config/schemas/config.yaml#L3263)), so I'm not sure what the standard is here?

Xartos commented on helmfile.d/values/admin-namespaces-wc.yaml.gotmpl:
```diff
@@ -70,6 +70,13 @@ namespaces:
       pod-security.kubernetes.io/warn: restricted
   {{ end }}
+  {{ if .Values.gpu.enabled }}
+  - name: gpu-operator
```
Question: Don't we need to add this namespace to more places? Like for some grafana dashboards or something.

Ajarmar replied:
I don't know, to be honest (haven't added a namespace before). I tried searching for other namespace names in the repo (like `hnc-system` and `fluentd-system`) and I don't see anything related to grafana dashboards? I did find that it should be added to `clean-wc.sh`, so I added it there.

Xartos commented on helmfile.d/stacks/gpu-operator.yaml.gotmpl:
```diff
@@ -0,0 +1,12 @@
+---
+templates:
+  nvidia-gpu-operator:
```
Question: This is probably me not remembering how this stacking works. But where do you specify that it's suppose to use the `upstream/nvidia/gpu-operator` chart?

Ajarmar replied:
[This file](https://github.com/elastisys/compliantkubernetes-apps/blob/2fda3c8587bb2140f9d083ffa914e657794b7d5c/helmfile.d/bases/upstream.yaml.gotmpl) generates templates for all upstream charts in the `upstream/index.yaml` file and then that template [is referred to here](https://github.com/elastisys/compliantkubernetes-apps/blob/2fda3c8587bb2140f9d083ffa914e657794b7d5c/helmfile.d/stacks/gpu-operator.yaml.gotmpl#L4).

Ajarmar commented:
I pushed some changes to change the name of the release to `nvidia-gpu-operator` to match the way we previously installed it, which should make migration simpler. Also added a migration script to migrate the `gpu-operatpr` namespace if it already exists. Output from testing the migration script:

```console
❯ CK8S_CLUSTER=wc ./20-migrate-gpu-operator-ns.sh execute
[ck8s] 20-migrate-gpu-operator-ns.sh: using config path: "/path/to/env"
[ck8s] 20-migrate-gpu-operator-ns.sh: using wc kubeconfig unencrypted
[ck8s] 20-migrate-gpu-operator-ns.sh: Migrating gpu-operator namespace to be managed by Helm if it already exists
[ck8s] 20-migrate-gpu-operator-ns.sh: Namespace gpu-operator exists, migrating namespace to be managed by Helm
namespace/gpu-operator labeled
namespace/gpu-operator annotated
namespace/gpu-operator annotated
Decrypting secret /path/to/env/secrets.yaml
Upgrading release=admin-namespaces, chart=charts/namespaces
Release "admin-namespaces" has been upgraded. Happy Helming!
NAME: admin-namespaces
LAST DEPLOYED: Fri Feb  7 16:05:03 2025
NAMESPACE: kube-system
STATUS: deployed
REVISION: 8
TEST SUITE: None

Listing releases matching ^admin-namespaces$
admin-namespaces	kube-system	8       	2025-02-07 16:05:03.940680104 +0100 CET	deployed	namespaces-0.1.1	           


UPDATED RELEASES:
NAME               CHART               VERSION   DURATION
admin-namespaces   charts/namespaces   0.1.1           4s
```

Ajarmar commented:
I've tested this on both CAPI and kubespray setups now.

aarnq commented on config/wc-config.yaml:
```diff
@@ -371,9 +371,10 @@ gpu:
-        - key: nvidia.com/gpu
-          operator: Exists
-          effect: NoSchedule
+        - key: "elastisys.io/node-type"
+          operator: "Equal"
+          value: "gpu"
+          effect: "NoSchedule"
```
nit: Remove quotes

aarnq commented on config/schemas/config.yaml:
```diff
@@ -7409,6 +7409,60 @@ properties:
+        description: Configure Node Feature Discovery
+        type: object
+        properties:
+          master:
+            title: Node Feature Discovery master Config
+            description: Configure Node Feature Discovery masters
```
Would have rather seen this be called `controlPlane`.

Ajarmar replied:
Sure. I updated it.

lunkan93 reviewed (approved):
LGTM

aarnq reviewed (approved):
Lgtm
